File: src-tauri/src/api/state/note_state.rs
//
// Copyright 2025 Shuntaro Kasatani
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

use knodiq_note::Note;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct NoteState {
    pub id: u32,
    pub pitch: u8,    // MIDI pitch
    pub velocity: u8, // MIDI velocity
    pub start_time: f32,
    pub duration: f32,
}

impl NoteState {
    pub fn from_note(note: &Note) -> Self {
        NoteState {
            id: note.id,
            pitch: note.pitch,
            velocity: note.velocity,
            start_time: note.start_beat,
            duration: note.duration,
        }
    }
}

impl Clone for NoteState {
    fn clone(&self) -> Self {
        NoteState {
            id: self.id,
            pitch: self.pitch,
            velocity: self.velocity,
            start_time: self.start_time,
            duration: self.duration,
        }
    }
}

File: src/lib/audio_api/note_state.ts
//
// Copyright 2025 Shuntaro Kasatani
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

export type NoteState = {
    /** The unique identifier for the note. */
    id: number;
    /** The pitch of the note, represented as a MIDI note number. */
    pitch: number;
    /** The velocity of the note, typically ranging from 0 to 127. */
    velocity: number;
    /** The start time of the note in beats. */
    start_time: number;
    /** The duration of the note in beats. */
    duration: number;
}